File: src/main/java/main/Function.java
package main;

import input.Group;
import input.Agent;
import org.jgrapht.alg.interfaces.ShortestPathAlgorithm;
import org.jgrapht.graph.DefaultUndirectedWeightedGraph;
import org.jgrapht.graph.DefaultWeightedEdge;

import java.util.HashMap;
import java.util.Set;

public class Function {
    HashMap<Integer, Agent> idToAgents;
    HashMap<Integer, Group> idToGroups;

    public Function(HashMap<Integer, Agent> idToRobots, HashMap<Integer, Group> idToGroups) {
        this.idToAgents = idToRobots;
        this.idToGroups = idToGroups;
    }

    public Double calculateOverLoadIS(Agent robot){
        //计算Individual Survivability
        double load = robot.getLoad();
        //获取组的生存评分
        double GS = calculateGS(idToGroups.get(robot.getGroupId()));
        //这里的存活率函数要好好设置一下
        return Math.max(GS*(1-sig(load/60)),0.3);

    }

    private Double calculateGS(Group group) {
        //计算Group Survivability
        double groupLoad = group.getGroupLoad();
        //用类似sigmod函数的变体做0-1之间的单调非增函数
        int size = group.getRobotIdInGroup().size();
        return Math.max(1-sig(groupLoad/(size*200)),0.6);
        //用来函数内进行调用
    }


    public Double sig(double x) {
        return (Math.exp(Math.log(x+1))-Math.exp(-Math.log(x+1)))/
                (Math.exp(Math.log(x+1))+Math.exp(-Math.log(x+1)));
    }

    public Double calculateContextualLoad(Agent leader, Agent robot,
                                          DefaultUndirectedWeightedGraph<Integer, DefaultWeightedEdge> arcGraph,
                                          ShortestPathAlgorithm<Integer, DefaultWeightedEdge> shortestPath,
                                          Double a , Double b) {
        double f = a*robot.getLoad()/robot.getCapacity()-b* calculateOverLoadIS(robot);
        //由相连边获取了domainF
        Set<DefaultWeightedEdge> defaultWeightedEdges = arcGraph.edgesOf(robot.getRobotId());
        double domianF = 0;
        double costSum = 0;
        for (DefaultWeightedEdge defaultWeightedEdge : defaultWeightedEdges) {
            Agent targetRobot = idToAgents.get(arcGraph.getEdgeTarget(defaultWeightedEdge));

            if (targetRobot.getGroupId()!=robot.getGroupId()||targetRobot.getRobotId()==robot.getRobotId()) {
                continue;
            }
            //与其相连边的沟通成本之和
            costSum+=arcGraph.getEdgeWeight(defaultWeightedEdge);
            domianF+=a*targetRobot.getLoad()/targetRobot.getCapacity()-b* calculateOverLoadIS(targetRobot);
        }
        int size = defaultWeightedEdges.size()+1;
        int domainNum = size +1;

        //加上进行层间任务迁移的cost
        double pathWeight = shortestPath.getPathWeight(leader.getRobotId(), robot.getRobotId());
        costSum+= pathWeight;
        //负载这个函数得想想
        return f+0.1*(domianF/domainNum+costSum/ size);
    }

}

File: python_src/main/function.py
"""Function class for calculating loads and survival rates"""
import math


class Function:
    def __init__(self, id_to_agents, id_to_groups):
        self.id_to_agents = id_to_agents
        self.id_to_groups = id_to_groups

    def calculate_over_load_is(self, robot):
        """Calculate Individual Survivability"""
        load = robot.get_load()
        # Get group survival score
        gs = self.calculate_gs(self.id_to_groups[robot.get_group_id()])
        # Survival rate function
        return max(gs * (1 - self.sig(load / 60)), 0.3)

    def calculate_gs(self, group):
        """Calculate Group Survivability"""
        group_load = group.get_group_load()
        # Use sigmoid-like function variant as monotonic decreasing function in 0-1
        size = len(group.get_robot_id_in_group())
        return max(1 - self.sig(group_load / (size * 200)), 0.6)

    def sig(self, x):
        """Sigmoid-like function"""
        return (math.exp(math.log(x + 1)) - math.exp(-math.log(x + 1))) / \
               (math.exp(math.log(x + 1)) + math.exp(-math.log(x + 1)))

    def calculate_contextual_load(self, leader, robot, arc_graph, shortest_path, a, b):
        """Calculate contextual load"""
        f = a * robot.get_load() / robot.get_capacity() - b * self.calculate_over_load_is(robot)

        # Get domain F from connected edges
        edges = arc_graph.edges(robot.get_robot_id())
        domain_f = 0.0
        cost_sum = 0.0

        for edge in edges:
            # Get the other endpoint of the edge
            if edge[0] == robot.get_robot_id():
                target_id = edge[1]
            else:
                target_id = edge[0]

            target_robot = self.id_to_agents[target_id]

            if target_robot.get_group_id() != robot.get_group_id() or \
               target_robot.get_robot_id() == robot.get_robot_id():
                continue

            # Sum of communication costs of connected edges
            cost_sum += arc_graph[edge[0]][edge[1]]['weight']
            domain_f += a * target_robot.get_load() / target_robot.get_capacity() - \
                       b * self.calculate_over_load_is(target_robot)

        size = len(list(edges)) + 1
        domain_num = size + 1

        # Add cost for inter-layer task migration
        try:
            path_weight = shortest_path(arc_graph, leader.get_robot_id(),
                                       robot.get_robot_id(), weight='weight')
        except:
            path_weight = 0

        cost_sum += path_weight

        return f + 0.1 * (domain_f / domain_num + cost_sum / size)
